r2-shared/src/main/java/org/readium/r2/shared/fetcher/ResourceInputStream.kt

```diff
@@ -15,35 +15,55 @@ import java.io.InputStream
 
 /** Input stream reading a [Resource]'s content. */
 class ResourceInputStream(
-    private val resource: Resource
+    private val resource: Resource,
+    val range: LongRange? = null
 ) : InputStream() {
 
-    private val length: Long by lazy {
-        try {
+    private var isClosed = false
+
+    private val end: Long by lazy {
+        val resourceLength = try {
             runBlocking { resource.length().getOrThrow() }
         } catch (e: Exception) {
             throw IOException("Can't get resource length")
         }
+
+        if (range == null)
+            resourceLength
+        else {
+            kotlin.math.min(resourceLength, range.last + 1)
+        }
+
     }
 
     /** Current position in the resource. */
-    private var position: Long = 0
+    private var position: Long = range?.start ?: 0
 
     /**
      * The currently marked position in the stream. Defaults to 0.
      */
     private var mark: Long = 0
```

Comment:
`mark` should probably be initialized like `position`, otherwise we can go back to the beginning of the resource using `reset()`.

```suggestion
    private var mark: Long = range?.start ?: 0
```

Reply:
Yes, thanks.